src/compiler/moduleSpecifiers.ts

```diff
@@ -844,17 +847,50 @@ const enum MatchingMode {
     Pattern,
 }
 
-function tryGetModuleNameFromExportsOrImports(options: CompilerOptions, targetFilePath: string, packageDirectory: string, packageName: string, exports: unknown, conditions: string[], mode: MatchingMode): { moduleFileToTry: string; } | undefined {
+function getOutputPath(targetFilePath: string, options: CompilerOptions, commonDir: string) {
+    if (options.emitDeclarationOnly) {
+        return;
+    }
+    return changeExtension(
+        options.outDir ?
+            resolvePath(
+                options.outDir,
+                getRelativePathFromDirectory(commonDir, targetFilePath, /*ignoreCase*/ false),
+            ) :
+            targetFilePath,
+        getOutputExtension(targetFilePath, options),
+    );
+}
+
+function tryGetModuleNameFromExportsOrImports(options: CompilerOptions, host: ModuleSpecifierResolutionHost, getCanonicalFileName: GetCanonicalFileName, targetFilePath: string, packageDirectory: string, packageName: string, exports: unknown, conditions: string[], mode: MatchingMode, isImports: boolean): { moduleFileToTry: string; } | undefined {
     if (typeof exports === "string") {
+        const currentDirectory = host.getCurrentDirectory();
+        const commonDir = getNormalizedAbsolutePath(

```

Comment:
`tryGetModuleNameFromExportsOrImports` may get called many times, but the common source directory never changes. Can we try to compute it only once and pass it into here?

Reply:
What do you think about adding this to `ModuleSpecifierResolutionHost`? Even though only `#imports` need this - we are in a function that is also used for handling `exports` and the lowest common ancestor of all calls to this is basically the root of all of this functionality here ;p

Reply:
Fine with me, if it doesn’t become annoying to update all hosts that implement it

Reply:
This went smoother than anticipated. It turns out that the only other implementer of that was already implementing `getCommonSourceDirectory` 😂 It also showed me that this is already available as `program.getCommonSourceDirectory()`. It uses slightly different arguments than I used but I don't think this should matter - the Program API is likely more correct/complete.